Online-Store

BondPV opened the pull request:
1. Task: [https://github.com/rolling-scopes-school/tasks/tree/master/tasks/online-store-team](url)
2. Screenshot:
![main-page](https://user-images.githubusercontent.com/108265593/211357985-324a9d27-92b6-48de-849d-6de571a036f7.png)
![product-page](https://user-images.githubusercontent.com/108265593/211357992-d14c078c-7029-4fb5-bcd5-fc9208d8dae8.png)
![cart-page](https://user-images.githubusercontent.com/108265593/211358001-458098ad-39ae-4687-9c28-8009492dd0f6.png)

3. Deploy: https://bondpv.github.io/online-store/
4. Done 09.01.2022 / deadline 10.01.2022
5. Score: 300 / 300
    Completed items
    ✔️ Product page with filters, 120 points
    ✔️ Cart page of selected products, 60 points
    ✔️ Modal checkout (purchase) window for selected products, 50 points
    ✔️ Product description page, 40 points
    ✔️ Header contains the number of products and the total amount, 20 points
    ✔️ Page 404, 10 points

BondPV committed:
init: start online store task

BondPV committed:
fead: add test file to structure

BondPV committed:
fix: name folder

BondPV committed:
feat: add base css and html

BondPV committed:
Merge pull request #2 from BondPV/Project-structure

Project structure

Alesia-V175 committed:
refactor: update project rules

Alesia-V175 committed:
Merge pull request #3 from BondPV/update-project-rules

refactor: update project rules

Alesia-V175 committed:
refactor: add additional attributes in html file

Alesia-V175 committed:
feat: add favicon

Alesia-V175 committed:
Merge pull request #4 from BondPV/page-structure-html/css

feat: update html footer and add favicon

BondPV committed:
feat: products database

BondPV committed:
feat: products database

BondPV committed:
Merge branch 'branch-bondpv' of https://github.com/BondPV/online-store into branch-bondpv

Alesia-V175 committed:
feat: add airbnb for typescript

Alesia-V175 committed:
feat: install husky and lint-staged

Alesia-V175 committed:
refactor: update husky settings

refactor: update husky settings

BondPV committed:
feat: added basic logic of module interaction

Alesia-V175 committed:
Merge pull request #5 from BondPV/branch-bondpv

Branch bondpv

Alesia-V175 committed:
refactor: update husky settings

BondPV committed:
Merge pull request #6 from BondPV/update-project-rules

Add additional settings for code style project

Alesia-V175 committed:
Add header style (#7)

* feat: add  part of header

* feat: update header

Alesia-V175 committed:
fix: update webpack config

BondPV committed:
Merge pull request #8 from BondPV/update-project-rules

fix: update webpack config

BondPV committed:
fix: project structure

BondPV committed:
fix: delete unnecessary rules in webpack.config file which deal with music and video

BondPV committed:
fix: project structure and project config

BondPV committed:
Merge pull request #9 from BondPV/update-project-structure

fix: project structure

Alesia-V175 committed:
feat: add stuly of catalog

Alesia-V175 committed:
feat: implement rating

Alesia-V175 committed:
refactor: add an external library in styles

Alesia-V175 committed:
Merge pull request #10 from BondPV/style-page

Style page of catalog

Alesia-V175 committed:
feat: add some code for sorting

BondPV committed:
feat: implement value filters

BondPV committed:
fix: value filters

BondPV committed:
fix: constants

Alesia-V175 committed:
feat: implement grid page, add 1 point of sorting

BondPV committed:
fix: feat: add rangeSliderControl class

BondPV committed:
feat: add rangeSliderControl class

BondPV committed:
fix: filters.ts

BondPV committed:
Merge pull request #11 from BondPV/create-value-filters

Create value filters

Alesia-V175 committed:
refactor: update grid catalog

Alesia-V175 committed:
feat: add sorting and grid of catalog; fix: resolve merge conflicts

BondPV committed:
fix: filter.ts

Alesia-V175 committed:
fix: update code style and add class names in separate object

BondPV committed:
feat: add productsQuantity to Filters.ts

BondPV committed:
feat: add productsQuantity to Filters.ts

BondPV committed:
feat: reset filters button

feat: add reset filters button

Alesia-V175 committed:
refactor: update code style accoding to requirements

Alesia-V175 committed:
feat: add styly for sorting

Alesia-V175 committed:
Merge pull request #12 from BondPV/implement-sorting

Implement sorting

BondPV committed:
feat: implement router

fix: temp commit 2

fix: temp commit

feat: add router

refactor: changed several files

fix: merge conflict

BondPV committed:
Merge pull request #13 from BondPV/create-value-filters

Create filters

BondPV committed:
feat: new database.json

BondPV committed:
Merge pull request #14 from BondPV/implement-router

Implement router

BondPV committed:
feat: add product page details

BondPV committed:
fix: productDetails

BondPV committed:
Merge pull request #15 from BondPV/product-details-page

add product page details

Alesia-V175 committed:
fix: fix title detailes name

Alesia-V175 committed:
feat: implement cart button in main page

Alesia-V175 committed:
refactor: update code style

Alesia-V175 committed:
feat: add sort and grid settings to local storage

Alesia-V175 committed:
feat: implement addition products to cart

Alesia-V175 committed:
refactor: update products on cartPage

Alesia-V175 committed:
feat: implement cart

BondPV committed:
feat: add url hash router

BondPV committed:
feat: add sort and view to url

BondPV committed:
feat: add copy link button

Alesia-V175 committed:
feat: implement cart

BondPV committed:
feat: search field

Alesia-V175 committed:
feat: implement promo block in cart page

BondPV committed:
feat: add errorPage and fix some code

BondPV committed:
fix: fixed filter with price and quantity

Alesia-V175 committed:
feat: update promo block and add styles

Alesia-V175 committed:
feat: implement pagination

Alesia-V175 committed:
feat: update pagination in cart page

Alesia-V175 committed:
refactor: update code style in cart.ts

BondPV committed:
fix: corrected remarks reviewers

BondPV committed:
Merge pull request #16 from BondPV/router-fix

Router fix

Alesia-V175 committed:
refactor: add class names to separate object

Alesia-V175 committed:
Merge branch 'develop' into implement-cart-page

# Conflicts:
#	src/components/main/catalog/Catalog.ts
#	src/components/main/grid/Grid.ts
#	src/components/main/product/Product.ts
#	src/components/main/product/product.scss
#	src/components/main/sortCatalog/SortCatalog.ts
#	src/components/pages/mainPage.ts
#	src/components/pages/productDetailsPage.ts
#	src/helpers/localStorage/LocalStorage.ts
#	src/index.html
#	src/types/enums.ts

Alesia-V175 committed:
fix: fix merge conflicts and fix comments

Alesia-V175 committed:
feat: create popup payment

BondPV committed:
add param cart to router

BondPV committed:
add param cart to router

BondPV committed:
fix: fixed styles

BondPV committed:
fix: fixed styles

Alesia-V175 committed:
Merge pull request #17 from BondPV/implement-cart-page

Implement cart page

BondPV committed:
fix: fixed search input

BondPV committed:
Merge branch 'add-param-cart-to-router' of https://github.com/BondPV/online-store into add-param-cart-to-router

BondPV committed:
Merge pull request #18 from BondPV/add-param-cart-to-router

Add param cart to router

BondPV committed:
feat: add self assessment

BondPV committed:
Merge pull request #19 from BondPV/self-assessment

feat: add self assessment

BondPV committed:
feat: add gh-pages - npm

Alesia-V175 committed:
feat: implement cart validation

Alesia-V175 committed:
Merge branch 'develop' into popup-payment

# Conflicts:
#	src/components/main/productDetails/ProductDetails.ts
#	src/components/pages/cartPage.ts
#	src/types/enums.ts

Alesia-V175 committed:
Merge pull request #21 from BondPV/popup-payment

Implement popup payment

Alesia-V175 committed:
feat: update cart render

Alesia-V175 committed:
Merge pull request #22 from BondPV/popup-payment

feat: update cart render

Alesia-V175 committed:
Update README.md

BondPV committed:
fix: fixed bug in links

BondPV committed:
feat: add jest tests settings

BondPV committed:
feat: add one test

BondPV committed:
Merge pull request #23 from BondPV/tests

Tests

Chichmarin17 commented on src/components/footer/footer.scss:
```diff
@@ -0,0 +1,42 @@
+.footer {
```
You are mixing styles with logic. In this case it would be better to move `header` and `footer` out of the `components` folder as it doesn't contain any business logic

Chichmarin17 commented on src/components/main/catalog/Catalog.ts:
```diff
@@ -0,0 +1,32 @@
+  render() {
+    const parentElement = document.querySelector(IdMap.catalog);
+    const products: HTMLElement[] = this.products
+      .map((product) => new Product(product))
+      .map((product) => product.render());
```
No need for an extra map, you can do it within the first one

Chichmarin17 commented on src/components/main/catalog/Catalog.ts:
```diff
@@ -0,0 +1,32 @@
+    }
+  }
+
+  notFound(): string {
```
What `notFound`? You can name it `getNotFoundElement` or `renderNotFoundElement`. Also, I can't see any usage of the method outside of the class, so it must be `private`

Chichmarin17 commented on src/components/main/payment/Payment.ts:
```diff
@@ -0,0 +1,97 @@
+                <span class="form__input_error" aria-live="polite"></span>
+              </div>
+              <div class="form__item">
+                <label for="email" class="form__label">E-mail</label>
+                <input id="email" type="email" placeholder="ivanIvanov@gmail.com" 
+                minlength = "5" maxlength="50" class="form__input" required>
+                <span class="form__input_error" aria-live="polite"></span>
+              </div>
+            </section>
+            <section>
+              <h3 class="popup__title">Credit card details</h3>
+              <div class="form__card">
+                <div class="form__item">
+                  <label for="card-number" class="form__label">Card Number</label>
+                  <span class="payment-icon"></span>
+                  <input id="card-number" type="number" placeholder="1234 5678 9012 3456" class="form__input" 
+                  pattern="[0-9]{4}"
+                  onkeypress="if(this.value.length === 16) return false;" required>
+                  <span class="form__input_error" aria-live="polite"></span>
+                </div>
+                <div class="item__wrap">
+                  <div class="form__item">
+                    <label for="card-valid" class="form__label">VALID</label>
+                    <input id="card-valid" type="text" placeholder="12/25" class="form__input" 
+                    onkeypress="if(this.value.length === 5) return false;" required>
+                    <span class="form__input_error" aria-live="polite"></span>
+                  </div>
+                  <div class="form__item"> 
+                    <label for="card-cvv" class="form__label">CVV</label>
+                    <input id="card-cvv" type="number" placeholder="123" class="form__input" 
+                    onkeypress="if(this.value.length === 3) return false;" required>
+                    <span class="form__input_error" aria-live="polite"></span>
+                  </div>
+                </div>
+              </div>
+            </section>
+            <button type="submit" class="form__button">CONFIRM</button>
+          </form>
+        </div>
+    `;
```
You can move it to `createPopupElement` private method or event move it to a separate class if there is any usage of the method outside of the current class

Chichmarin17 commented on src/components/main/payment/PaymentValidation.ts:
```diff
@@ -0,0 +1,244 @@
+
+    const inputFormArr = [nameUser, phone, address, email, cardCvv, cardNum, cardValid];
+
+    PaymentValidation.check(nameUser, Validation.NameInvalidMessage, (): boolean => {
+      const nameValue = nameUser.value.trim();
+      const splitedName = nameValue.split(' ');
+
+      if (nameUser.validity.valueMissing || nameUser.validity.typeMismatch || nameUser.validity.tooShort) {
+        return false;
+      }
+
+      if (splitedName.length < 2) {
+        return false;
+      }
+
+      for (let i = 0; i < splitedName.length; i++) {
+        if (splitedName[i].length < 3) {
+          return false;
+        }
+      }
```
You can move all the validation methods into separate methods

Chichmarin17 commented on src/components/main/payment/PaymentValidation.ts:
```diff
@@ -0,0 +1,244 @@
+      });
+
+      if (invalidInputs.length > 0) {
+        invalidInputs.forEach((input: HTMLElement) =>
+          PaymentValidation.showMessageError(input, Validation.ErrorMessage),
+        );
+      } else {
+        const bodyContainer = document.querySelector('.body');
+        let seconds = 5;
+
+        const popup = document.querySelector('.popup');
+        const popupOverlay = document.querySelector('.popup__overlay');
+
+        if (popup) {
+          popup.remove();
+        }
+
+        if (bodyContainer && popupOverlay && bodyContainer) {
+          const closePopup = document.createElement('div');
+          closePopup.classList.add('popup_close');
+          closePopup.textContent = `Thank you for your order. Redirect to the store after ${seconds} sec.`;
+          bodyContainer.append(closePopup);
+
+          const showText = setInterval(() => {
+            seconds -= 1;
+            closePopup.textContent = `Thanks for your order. Redirect to the store after ${seconds} sec.`;
+          }, 1000);
+
+          setTimeout(() => {
+            closePopup.remove();
+            popupOverlay.remove();
+            bodyContainer.classList.remove('body_active');
+            clearInterval(showText);
+            LocalStorage.removeProductsFromCart();
+            window.location.hash = `#${Pages.Main}`;
+          }, 5000);
+        }
+      }
+      event.preventDefault();
+    });
```
Too much logic in the event listener callback. Split it into multiple methods as `validateSmth` & `createRedirectionTimer`

Chichmarin17 commented on src/components/main/product/Product.ts:
```diff
@@ -0,0 +1,122 @@
+    const ratingStars: string = Product.fillRating(this.product.rating);
+    const initialPrice: number = this.initialPrice();
+
+    const productString = `
+      <div class="product">
+        <div class="product__wrap">
+          <div class="product__stock">stock: ${this.product.stock}</div>
+          <div class="product__descriptionk">category: ${this.product.category}</div>
+          <div class="product__img-wrap">
+            <a href="#product-details/${this.product.id}">
+              <div class="product__img" style="background: 
+              url('${this.product.thumbnail}') no-repeat center / contain;"></div>
+            </a>
+          </div>
+        </div>
+        <div class="product__wrap">
+          <div class="product__title">${this.product.title}</div>
+          <div class="product__rating-wrap">
+            <div class="product__rating">${ratingStars}</div>
+            <span class="product__rating-text">${this.product.rating}</span>
+          </div>
+          <div class="product__description">${this.product.titleDetail}</div>
+          <div class="product__footer-wrap">
+            <div class="product__price-wrap">
+              <div class="product__price">${Symbol.Currence} ${initialPrice}</div>
+              <div class="product__price_discount">${Symbol.Currence} ${this.product.price}</div>
+            </div>
+          </div>
+        </div> 
+      </div>
+      `;
```
Move it out

Chichmarin17 commented on src/components/main/productDetails/ProductDetails.ts:
```diff
@@ -0,0 +1,153 @@
+import { Pages, Symbol } from 'types/enums';
+
+class ProductDetails extends Product {
+  private price(): HTMLElement[] {
```
What `price`? Check what does function return and name it like `getPrice`

Chichmarin17 commented on src/components/main/productDetails/ProductDetails.ts:
```diff
@@ -0,0 +1,153 @@
+    return [productPrice, productPriceDiscount];
+  }
+
+  public renderHeader(): HTMLElement {
```
Could be better to move it to a separate component (you have `header` folder)

Chichmarin17 commented on src/components/main/productDetails/ProductDetails.ts:
```diff
@@ -0,0 +1,153 @@
+
+  private addToCart(buttonAddToCart: HTMLElement): void {
+    if (buttonAddToCart.classList.length === 1) {
+      buttonAddToCart.textContent = 'REMOVE FROM CART';
+    } else {
+      buttonAddToCart.textContent = 'ADD TO CART';
```
Better to move all magic numbers/strings into consts. What if you want to use those values again? you ought to search for all mentions of the same string in different places. It could easily lead to multiple minor bugs

Chichmarin17 commented on src/components/main/productDetails/ProductDetails.ts:
```diff
@@ -0,0 +1,153 @@
+    const productGallery: HTMLElement = document.createElement('div');
+    productGallery.classList.add('product-details__gallery');
+
+    new Gallery(productGallery, this.product.images);
```
What is the usage of instantiating a class without assigning it to any variable? I guess you move all logic of the `Gallery` class inside its constructor. That's completely wrong, you should better create a method for combining all the class actions (like `render` or `start`)

Chichmarin17 commented on src/components/main/sortBar/filters/copyLinkButton/CopyLinkButton.ts:
```diff
@@ -0,0 +1,23 @@
+    copyButton.addEventListener('click', () => {
+      navigator.clipboard.writeText(window.location.href);
+      copyButton.innerText = 'Copied!';
+      setTimeout(() => {
+        copyButton.innerText = 'Copy Link';
+      }, 300);
```
I saw `setTimeout` before with a different value. You should get rid of those magical numbers by moving those into a separate file

Chichmarin17 commented on src/components/pages/errorPage/errorPage.ts:
```diff
@@ -0,0 +1,26 @@
+import { Container } from 'constants/htmlConstants';
+import './errorPage.scss';
+
+class ErrorPage {
+  constructor(private container: HTMLElement = Container) {
+    this.renderPage();
+  }
+
+  public renderPage(): void {
+    const errorPageContainer: HTMLElement = document.createElement('div');
+    errorPageContainer.classList.add('error');
+    errorPageContainer.innerHTML = `
+      <div class="error__descr">
+        <p>Error 404!</p>
+        <span>Something went wrong...</span>
+      </div>
+      `;
+    this.container.append(errorPageContainer);
+  }
+
+  public removePage(): void {
+    this.container.innerHTML = '';
+  }
+}
```
Pretty weird class. It contains `render` method (it's public) but at the same time it calls the method inside a constructor. Just call the method from the instance, your constructor should not incapsulate your public methods

Chichmarin17 commented on src/constants/htmlConstants.ts:
```diff
@@ -0,0 +1,117 @@
+
+export const valueOptionAsc = 'ASC';
+
+export const fullStarIcon = '<i class="fa-solid fa-star rating-star"></i>';
+
+export const halfStarIcon = '<i class="fa-regular fa-star-half-stroke rating-star"></i>';
```
You mixed JS consts with HTML, better split those up

Chichmarin17 commented on src/helpers/router/UrlHash.ts:
```diff
@@ -0,0 +1,107 @@
+    const sort = `${FiltersName.Sort}=${hashData.sort}`;
+    const view = `${FiltersName.View}=${hashData.view}`;
+
+    window.location.hash = `?${category}&${brand}&${price}&${stock}&${search}&${sort}&${view}`;
```
What if you have an empty `brand` filter, it will handle that case?

Chichmarin17 commented on src/helpers/router/UrlHash.ts:
```diff
@@ -0,0 +1,107 @@
+    return this.hashData[param];
+  }
+
+  public static setUrlHashParam(param: FiltersName.Sort | FiltersName.Search | FiltersName.View, value: string): void {
+    this.hashData[param] = value;
+
+    if (param !== FiltersName.Sort || (param === FiltersName.Sort && value !== '')) {
+      this.setUrlHash(this.hashData);
+    }
+  }
+
+  static clearHash() {
```
What is the difference between `static` and `public static`? public method could be called from the class instance, static method is callable directly from the class itself

BondPV committed:
Delete .idea directory

BondPV committed:
Delete .husky directory

dependabot[bot] committed:
build(deps): bump json5 from 1.0.1 to 1.0.2

Bumps [json5](https://github.com/json5/json5) from 1.0.1 to 1.0.2.
- [Release notes](https://github.com/json5/json5/releases)
- [Changelog](https://github.com/json5/json5/blob/main/CHANGELOG.md)
- [Commits](https://github.com/json5/json5/compare/v1.0.1...v1.0.2)

---
updated-dependencies:
- dependency-name: json5
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

BondPV committed:
Merge pull request #24 from BondPV/dependabot/npm_and_yarn/json5-1.0.2

build(deps): bump json5 from 1.0.1 to 1.0.2
